File: tests/houses_stake/add_reward.ts
import * as anchor from '@coral-xyz/anchor';
import { Program, BN } from '@coral-xyz/anchor';
import { HousesStake } from '../../target/types/houses_stake';
import { requestAirdrop } from '../../helpers/requestAirdrop';
import {
  ASSOCIATED_TOKEN_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
  getAccount,
  mintTo,
} from '@solana/spl-token';
import { getLocalRewardMint } from '../../helpers/getLocalRewardMint';
import { assert } from 'chai';
import { getStakeAccounts } from '../../helpers/houses_stake/getStakeAccounts';

export const callAddReward = async (amount) => {
  const rewardMint = getLocalRewardMint();
  const {
    adminAccount,
    dataPda,
    program,
    authorityTokenAccount,
    rewardTokenAccount,
    connection,
    mintAuthority,
  } = await getStakeAccounts();
  await requestAirdrop(connection, adminAccount.publicKey, 5 * 10e5);
  await mintTo(
    connection,
    adminAccount,
    rewardMint,
    authorityTokenAccount.address,
    mintAuthority,
    1500
  );

  await program.methods
    .addReward(new BN(amount))
    .accounts({
      systemProgram: anchor.web3.SystemProgram.programId,
      tokenProgram: TOKEN_PROGRAM_ID,
      associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
      dataPda,
      rewardTokenAccount,
      authorityTokenAccount: authorityTokenAccount.address,
      authority: adminAccount.publicKey,
    })
    .signers([adminAccount])
    .rpc();
};

describe('add reward', () => {
  const program = anchor.workspace.HousesStake as Program<HousesStake>;
  anchor.setProvider(anchor.AnchorProvider.env());
  const connection = program.provider.connection;

  it('add reward', async () => {
    const { dataPda, rewardTokenAccount } = await getStakeAccounts();

    let dataAccount = await program.account.data.fetch(dataPda);
    const rewardIndexBefore = dataAccount.rewardIndex;
    const rewardAmountBalanceBefore = (
      await getAccount(connection, rewardTokenAccount)
    ).amount;
    let { stacked: stackedBefore } = dataAccount;
    await callAddReward(1000);
    const rewardAmountBalance = (
      await getAccount(connection, rewardTokenAccount)
    ).amount;
    assert.equal(rewardAmountBalance, rewardAmountBalanceBefore + BigInt(1000));
    const { rewardIndex, stacked, previousStacked } =
      await program.account.data.fetch(dataPda);
    assert.equal(rewardIndexBefore + 1, rewardIndex);
    assert.equal(0, stacked.toNumber());
    assert.equal(stackedBefore.toNumber(), previousStacked.toNumber());
  });
});

File: programs/houses_stake/src/methods/add_reward.rs
use anchor_lang::prelude::*;
use anchor_spl::{token::{Token, TokenAccount, self}, associated_token::AssociatedToken};
use crate::types::*;

pub fn add_reward(ctx: Context<AddReward>, amount: u64) -> Result<()> {
    let cpi_ctx: CpiContext<token::Transfer> = CpiContext::new(
        ctx.accounts.token_program.to_account_info(),
        token::Transfer {
            from: ctx.accounts.authority_token_account.to_account_info(),
            authority: ctx.accounts.authority.to_account_info(),
            to: ctx.accounts.reward_token_account.to_account_info(),
        },
    );
    token::transfer(cpi_ctx, amount)?;
    ctx.accounts.data_pda.reward_index = (ctx.accounts.data_pda.reward_index + 1) % 255;
    ctx.accounts.data_pda.reward = amount;
    ctx.accounts.data_pda.previous_stacked = ctx.accounts.data_pda.stacked;
    ctx.accounts.data_pda.stacked = 0;
    Ok(())
}

#[derive(Accounts)]
#[instruction(amount: u64)]
pub struct AddReward<'info> {
    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,

    #[account(mut,
        seeds = [ b"data".as_ref(), authority.key.as_ref() ],
        // add constraint for freeze staking
        bump)]
    pub data_pda: Account<'info, Data>,

    #[account(
        mut,
        associated_token::mint = REWARD_MINT_ADDRESS.parse::<Pubkey>().unwrap(),
        associated_token::authority = data_pda
    )]
    pub reward_token_account: Account<'info, TokenAccount>,

    #[account(mut,
        associated_token::mint = REWARD_MINT_ADDRESS.parse::<Pubkey>().unwrap(),
        constraint = authority_token_account.amount >= amount,
        associated_token::authority = authority)]
    pub authority_token_account: Account<'info, TokenAccount>,

    #[account(address = AUTHORITY_ADDRESS.parse::<Pubkey>().unwrap())]
    pub authority: Signer<'info>,
}
